source/server/admin/stats_html_render.cc

```diff
@@ -0,0 +1,188 @@
+}
+
+void StatsHtmlRender::finalize(Buffer::Instance& response) {
+  ASSERT(!finalized_);
+  finalized_ = true;
+  if (has_pre_) {
+    response.add("</pre>\n");
+  }
+  response.add("</body>\n");
+}
+
+void StatsHtmlRender::startPre(Buffer::Instance& response) {
+  has_pre_ = true;
+  response.add("<pre>\n");
+}
+
+void StatsHtmlRender::generate(Buffer::Instance& response, const std::string& name,
+                               const std::string& value) {
+  response.addFragments({name, ": \"", Html::Utility::sanitize(value), "\"\n"});
+}
+
+void StatsHtmlRender::noStats(Buffer::Instance& response, absl::string_view types) {
+  response.addFragments({"</pre>\n<br/><i>No ", types, " found</i><br/>\n<pre>\n"});
+}
+
+void StatsHtmlRender::tableBegin(Buffer::Instance& response) { response.add(AdminHtmlTableBegin); }
+
+void StatsHtmlRender::tableEnd(Buffer::Instance& response) { response.add(AdminHtmlTableEnd); }
+
+void StatsHtmlRender::urlHandler(Buffer::Instance& response, const Admin::UrlHandler& handler,
+                                 OptRef<const Http::Utility::QueryParams> query) {
+  absl::string_view path = handler.prefix_;
+
+  if (path == "/") {
+    return; // No need to print self-link to index page.
+  }
+
+  // Remove the leading slash from the link, so that the admin page can be
+  // rendered as part of another console, on a sub-path.
+  //
+  // E.g. consider a downstream dashboard that embeds the Envoy admin console.
+  // In that case, the "/stats" endpoint would be at
+  // https://DASHBOARD/envoy_admin/stats. If the links we present on the home
+  // page are absolute (e.g. "/stats") they won't work in the context of the
+  // dashboard. Removing the leading slash, they will work properly in both
+  // the raw admin console and when embedded in another page and URL
+  // hierarchy.
+  ASSERT(!path.empty());
+  ASSERT(path[0] == '/');
+  std::string sanitized_path = Html::Utility::sanitize(path.substr(1));
+  path = sanitized_path;
+
+  // Alternate gray and white param-blocks. The pure CSS way of coloring based
+  // on row index doesn't work correctly for us we are using a row for each
+  // parameter, and we want each endpoint/option-block to be colored the same.
+  const char* row_class = (++index_ & 1) ? " class='gray'" : "";
+
+  // For handlers that mutate state, render the link as a button in a POST form,
+  // rather than an anchor tag. This should discourage crawlers that find the /
+  // page from accidentally mutating all the server state by GETting all the hrefs.
+  const char* method = handler.mutates_server_state_ ? "post" : "get";
+  if (submit_on_change_) {
+    response.addFragments({"\n<form action='", path, "' method='", method, "' id='", path,
+                           "' class='home-form'></form>\n"});
+  } else {
+    // Render an explicit visible submit as a link (for GET) or button (for POST).
+    const char* button_style = handler.mutates_server_state_ ? "" : " class='button-as-link'";
```

Comment:
(Note this will also allow us to remove things like "pointer: cursor" and things that shouldn't apply to links in terms of styling)

Reply:
The reason I think it needs to be a form is that it needs to collect some input parameters. This page is a little more complex as it shows distinct form parameters for each endpoint. The endpoints that mutate state are rendered as buttons and issue POSTs. The endpoints that just return info are rendered as links and issue GETs but they both need to collect their associated parameters.

In the screen-shot in the description you can see a POST example "drain_listeners" and you can see a GET example "config_dump".

I can also attach the generated HTML for this if that helps.

Reply:
Oh I think I see your point -- many of the endpoints do not have parameters and they could be simple links. I'd be OK with that if that makes the screen-reader experience better.

What's the easiest way for me to see that that looks like with a screen reader?

Reply:
Yeah I'm purely talking about the ones without parameters here (the ones with it should ideally be so too, but that requires JS which is a much longer discussion I don't think we wanna take on, plus having the page work without JS is nice). Sorry if that wasn't clear!

As for testing with a screenreader yourself it depends a bit on the OS, if you're using a Mac it's just in settings as voiceover, and they should also have a tutorial. On windows there isn't a good one out of the box , you'd want to use NVDA (since Jaws the other popular one isn't free), though there's no tutorial. Linux has ORCA, but that's requires lots of finagling in my experience. 

Screenreader testing is also usually very context dependent as it can be hard to know how something should be "read". I'd be happy to discuss this in slack, or over a call, but the core things this helps with are:

1. Most screenreaders have a concept of a "rotor" which is for quickly navigating landmarks/forms/etc. of a page, the more helpful we can make this rotor by say not putting a bunch of forms on, the easier it is to quickly navigate.
2. On Windows, screenreaders have separate modes depending on if you're in a "form" or not. Many users switch to form mode whenever they hear they're entering a fork to better handle it. To get presented a form with no content is quite confusing when you can't see there's no data to input.

Reply:
Thanks -- I think I did the corrections you suggested, though the exact structure you specified was not correct according an HTML validator site. I corrected that and tried the Mac screen reader, though I couldn't detect any difference in the behavior; it seemed to work OK.

In the form that's committed to the PR now, I get only one validation error for the inline favicon, though I think the validator is wrong in that case; the favicon works in Firefox. It doesn't show up in Chrome for me because Chrome overrides the favicon for HTTP sites as a warning to users that the connection is not private.

PTAL.

Reply:
I have updated the GET endpoints without parameters as links. I  am not 100% sure this change makes things better.

If it does, should we also render the POST endpoints without parameters without a form? Maybe a simple button?

Reply:
Hey thanks for doing this! I can take a look here, I imagine it'll improve the rotor for OSX, and the Windows "Form" vs "Non-Form" mode. I'm happy to take a peek with both.

As for "POST" parameters, we'd either need to do a form or something with javascript to send a post on button click. I don't think we need to complicate it at this stage by adding JS to the site, should we make it in the long term? Maybe. I wouldn't worry about that for now.